File: expo-maps/android/src/main/java/GoogleMapsMarkers.kt
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import expo.modules.maps.MarkerObject
import expo.modules.maps.Markers

class GoogleMapsMarkers(map: GoogleMap): Markers {
    private val markers = mutableListOf<Marker>()
    private var googleMap: GoogleMap = map

    override fun setMarkers(markerObjects: Array<MarkerObject>) {
        this.detachAndDeleteMarkers()
        for (markerObject in markerObjects) {
            val markerOptions = MarkerOptions()
            markerOptions.position(LatLng(markerObject.latitude!!, markerObject.longitude!!))
            val marker = this.googleMap.addMarker(markerOptions)
            this.markers.add(marker!!)
        }
    }

    override fun detachAndDeleteMarkers() {
        for (marker in this.markers) {
            marker.remove()
        }
        this.markers.clear()
    }
}

File: expo-maps/ios/ExpoMaps/GoogleMaps/GoogleMapsMarkers.swift
import GoogleMaps

class GoogleMapsMarkers: Markers {

  private let mapView: GMSMapView
  private var markers: [GMSMarker] = []
  private var poiMarkers: [GMSMarker] = []

  init(mapView: GMSMapView) {
    self.mapView = mapView
  }

  func setMarkers(markerObjects: [MarkerObject]) {
    detachAndDeleteMarkers()
    for markerObject in markerObjects {
      let marker: GMSMarker = createGoogleMarker(markerObject: markerObject)
      
      marker.map = mapView
      markers.append(marker)
    }
  }
  
  func setPOIMarkers(markerObjects: [MarkerObject]) {
    detachAndDeletePOIMarkers()
    for markerObject in markerObjects {
      let marker: GMSMarker = createGoogleMarker(markerObject: markerObject)
      
      marker.map = mapView
      poiMarkers.append(marker)
    }
  }

  internal func detachAndDeleteMarkers() {
    for marker in markers {
      marker.map = nil
    }
    markers = []
  }
  
  func detachAndDeletePOIMarkers() {
    for marker in poiMarkers {
      marker.map = nil
    }
    poiMarkers = []
  }
}
